prov/lnx/src/lnx_cq.c

```diff
@@ -0,0 +1,241 @@
+		if (rc)
+			frc = rc;
+		ep->lpe_cq.lpc_core_cq = NULL;
+	}
+
+	return frc;
+}
+
+static int lnx_cq_close(struct fid *fid)
+{
+	int rc;
+	struct lnx_cq *lnx_cq;
+	struct local_prov *entry;
+	struct dlist_entry *prov_table;
+
+	lnx_cq = container_of(fid, struct lnx_cq, util_cq.cq_fid);
+	prov_table = &lnx_cq->lnx_domain->ld_fabric->local_prov_table;
+
+	/* close all the open core cqs */
+	dlist_foreach_container(prov_table, struct local_prov,
+				entry, lpv_entry) {
+		rc = lnx_cleanup_cqs(entry);
+		if (rc) {
+			FI_INFO(&lnx_prov, FI_LOG_CORE, "Failed to close domain for %s\n",
+				entry->lpv_prov_name);
+			return rc;
+		}
+	}
+
+	rc = ofi_cq_cleanup(&lnx_cq->util_cq);
+	if (rc)
+		return rc;
+
+	free(lnx_cq);
+	return 0;
+}
+
+struct fi_ops_cq_owner lnx_cq_write = {
+	.size = sizeof(lnx_cq_write),
+	.write = lnx_peer_cq_write,
+	.writeerr = lnx_peer_cq_writeerr,
+};
+
+static struct fi_ops lnx_cq_fi_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = lnx_cq_close,
+	.bind = fi_no_bind,
+	.control = ofi_cq_control,
+	.ops_open = fi_no_ops_open,
+};
+
+static void lnx_cq_progress(struct util_cq *cq)
+{
+	struct lnx_cq *lnx_cq;
+	struct local_prov_ep *ep;
+	struct local_prov *entry;
+	struct dlist_entry *prov_table;
+
+	lnx_cq = container_of(cq, struct lnx_cq, util_cq);
+	prov_table = &lnx_cq->lnx_domain->ld_fabric->local_prov_table;
+/*
+	entry = lnx_cq->lnx_domain->ld_fabric->shm_prov;
+	dlist_foreach_container(&entry->lpv_prov_eps,
+				struct local_prov_ep, ep, entry)
+		fi_cq_read(ep->lpe_cq.lpc_core_cq, NULL, 0);
+
+	return;
+*/
+	/* Kick the core provider endpoints to progress */
+	dlist_foreach_container(prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+					struct local_prov_ep, ep, entry)
+			fi_cq_read(ep->lpe_cq.lpc_core_cq, NULL, 0);
+	}
+}
+
+static int lnx_cq_open_core_prov(struct lnx_cq *cq, struct fi_cq_attr *attr)
+{
+	int rc;
+	struct local_prov_ep *ep;
+	struct local_prov *entry;
+	struct dlist_entry *prov_table =
+		&cq->lnx_domain->ld_fabric->local_prov_table;
+
+	/* tell the core providers to import my CQ */
+	attr->flags |= FI_PEER;
+
+	/* create all the core provider endpoints */
+	dlist_foreach_container(prov_table, struct local_prov,
+				entry, lpv_entry) {
+		dlist_foreach_container(&entry->lpv_prov_eps,
+					struct local_prov_ep, ep, entry) {
+			struct fid_cq *core_cq;
+			struct fi_peer_cq_context cq_ctxt;
+
+			ep->lpe_cq.lpc_shared_cq = cq;
+			ep->lpe_cq.lpc_cq.owner_ops = &lnx_cq_write;
+
+			cq_ctxt.size = sizeof(cq_ctxt);
+			cq_ctxt.cq = &ep->lpe_cq.lpc_cq;
+
+			/* pass my CQ into the open and get back the core's cq */
+			rc = fi_cq_open(ep->lpe_domain, attr, &core_cq, &cq_ctxt);
+			if (rc)
+				return rc;
+
+			/* before the fi_cq_open() returns the core provider should
+			 * have called fi_export_fid() and got a pointer to the peer
+			 * CQ which we have allocated for this core provider
+			 */
+
+			ep->lpe_cq.lpc_core_cq = core_cq;
```

Comment:
the logic of cq creation and binding is confusing.  The core_cq's are supposed to be part of the lnx_cq. The reference to the cq in the ep should be set when the cq is bound to the ep, not now.

same issue exists for other objects such as av.